### ISSUE TITLE
`storeDir` directive fails in some cases when using `google-lifesciences` executor 

### ISSUE THREAD

**Issue by spvensko** [nextflow-io/nextflow#2362]
### Expected behavior and actual behavior

I am using `storeDir` with the `google-lifesciences` executor. I expect for a process with a `storeDir` directive ran using `google-lifesciences` to copy the captured output from the output block to the local path defined within the `storeDir` directive definition within the process. `publishDir`, when defined for a process and ran through `google-lifesciences`, is able to copy the output from the cloud gs:// bucket work directory to the local path defined within the `publishDir` directive definition.

### Steps to reproduce the problem

Consider the following script:

**minimal_example.nf**
```
nextflow.enable.dsl=2

include { minimal_process } from './minimal_module/minimal_module.nf'

workflow {
  minimal_process()
}
```
and the following module:

**./minimal_module/minimal_module.nf**
```
process minimal_process {


  container 'ubuntu:xenial'
  publishDir "/home/spvensko/minimal_example_outdir/publish_dir/"
  storeDir "/home/spvensko/minimal_example_outdir/store_dir/"

  output:
  path("this_is_an_example_file"), emit: example_file

  script:
  """
  touch this_is_an_example_file
  """
}
```

Ran the following ways:

Using `local`:
```
nextflow minimal_example.nf
```

Using `slurm`:
```
nextflow minimal_example.nf -profile lbg_slurm
```

Using `google-lifescienes` (with `publishDir` *and* `storeDir` directives:
```
nextflow minimal_example.nf` -profile -lbg_gcp -bucket-dir gs://spvensko/work
```

Using `google-lifesciences` (with *only* `publishDir`)
```
nextflow minimal_example.nf` -profile -lbg_gcp -bucket-dir gs://spvensko/work
```

### Program output 

Note: I am clearing the output directory each time prior to running. I am also including the `ls` outputs to show that the files are being copied as expected in the passing cases.

Using `local`:
```
[spvensko@c6145-2-8 ~]$ nextflow minimal_example.nf 
N E X T F L O W  ~  version 21.08.0-edge
Launching `minimal_example.nf` [cranky_stonebraker] - revision: 4df81d0fdf
executor >  local (1)
[fa/9743cc] process > minimal_process [100%] 1 of 1 ✔
[spvensko@c6145-2-8 ~]$ ls -lhdrt minimal_example_outdir/*/*
-rw-r--r--. 1 spvensko lbg  0 Oct  1 14:45 minimal_example_outdir/store_dir/this_is_an_example_file
lrwxrwxrwx. 1 spvensko lbg 71 Oct  1 14:45 minimal_example_outdir/publish_dir/this_is_an_example_file -> /home/spvensko/minimal_example_outdir/store_dir/this_is_an_example_file
```

Using `slurm`:
```
[spvensko@c6145-2-8 ~]$ nextflow minimal_example.nf -profile lbg_slurm
N E X T F L O W  ~  version 21.08.0-edge
Launching `minimal_example.nf` [stupefied_feynman] - revision: 4df81d0fdf
executor >  slurm (1)
[44/45395d] process > minimal_process [100%] 1 of 1 ✔
Completed at: 01-Oct-2021 15:30:47
Duration    : 40m 41s
CPU hours   : (a few seconds)
Succeeded   : 1
[spvensko@c6145-2-8 ~]$ ls -lhdrt minimal_example_outdir/*/*
-rw-r--r--. 1 spvensko lbg  0 Oct  1 15:30 minimal_example_outdir/store_dir/this_is_an_example_file
lrwxrwxrwx. 1 spvensko lbg 71 Oct  1 15:30 minimal_example_outdir/publish_dir/this_is_an_example_file -> /home/spvensko/minimal_example_outdir/store_dir/this_is_an_example_file
```

Using `google-lifesciences` (with `publishDir` *and* `storeDir` directives):
```
(raft-test) [spvensko@c6145-2-8 ~]$ nextflow minimal_example.nf -profile lbg_gcp -bucket-dir gs://spvensko/work              N E X T F L O W  ~  version 21.08.0-edge
Launching `minimal_example.nf` [lonely_northcutt] - revision: 4df81d0fdf
executor >  google-lifesciences (1)
[18/2567c4] process > minimal_process [100%] 1 of 1, failed: 1 ✘
Error executing process > 'minimal_process'
Caused by:
  Missing output file(s) `this_is_an_example_file` expected by process `minimal_process`
Command executed:
  touch this_is_an_example_file
Command exit status:
  0
Command output:
  (empty)
Work dir:
  gs://spvensko/work/18/2567c4de99c9a1ea3f1ef428e1ad37
Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`
[spvensko@c6145-2-8 ~]$ ls -lhdrt minimal_example_outdir/*/*
ls: cannot access minimal_example_outdir/*/*: No such file or directory
```

Using `google-lifesciences` (with *only* publishDir):
```
spvensko@c6145-2-8 ~]$ nextflow minimal_example.nf -profile lbg_gcp -bucket-dir gs://spvensko/work
N E X T F L O W  ~  version 21.08.0-edge
Launching `minimal_example.nf` [determined_goldwasser] - revision: 4df81d0fdf
executor >  google-lifesciences (1)
[55/cb0b59] process > minimal_process [100%] 1 of 1 ✔
Completed at: 01-Oct-2021 16:02:25
Duration    : 1m 41s
CPU hours   : (a few seconds)
Succeeded   : 1
[spvensko@c6145-2-8 ~]$ ls -lhdrt minimal_example_outdir/*/*
-rw-r--r--. 1 spvensko lbg 0 Oct  1 16:02 minimal_example_outdir/publish_dir/this_is_an_example_file
```

### Environment 

* Nextflow version: `21.08.0-edge` (though relevant to other versions as well)
* Java version: `1.8.0_222-ea`
* Operating system: host: `CentOS Linux 7`; `google-lifesciences` Docker container: `ubuntu:xenial`
* Bash version: host: `4.2.46(2)`

### Additional context

`publishDir` is able to copy from the gs:// bucket work directory to my local directory, but `storeDir`, under the same conditions, cannot. This appears to be due to how `publishDir` is ran (asynchronously in the background by Nextflow, _not_ explicitly defined within the process's `.command.run` file). `storeDir`, on the other hand, is explicitly defined within the `nxf_unstage()` definition:

**work_18_2567c4de99c9a1ea3f1ef428e1ad37_.command.run**
```
    ...
    uploads=()
    IFS=$'\n'
    for name in $(eval "ls -1d this_is_an_example_file" | sort | uniq); do
        uploads+=("nxf_gs_upload '$name' /home/spvensko/minimal_example_outdir/store_dir")
    done
    unset IFS
    nxf_parallel "${uploads[@]}"
    ....
```

Also, it may be worth noting that the `nxf_gs_upload` command appears to fail which results in the contents of the work directory (except for `.command.*` and `.exitcode`) being deleted. I suspect this is expected behavior, but it the source of the `Missing output file(s)` error. 

It's not clear to me why `storeDir` is defined one way while `publishDir` is defined another way. The only advantage I could see if that the outputs moved by `storeDir` would be immediately available for usage while `publishDir` outputs would not (hence the `publishDir` asynchronous copying warning in the documentation).

In short, would it be possible to allow `storeDir` directive to have an alternate mode in which it operates *outside* of the `.command.run` definition in the same way `publishDir` does? This would allow compatibility with the `google-lifesciences` executor when using it in the "hybrid" mode but also allow for the benefits of `storeDir` when it comes to storing previously generated outputs.

**Comment by bentsherman** [comment]
Possibly solved by #3016
